marshal/cluster.go

```diff
@@ -0,0 +1,272 @@
+	// queries based on THIS time instead of the current, actual time.
+	ts int64
+}
+
+// getOrCreateCluster will either return a pointer to a valid cluster object if it exists or
+// create a new one and then return it.
+func getOrCreateCluster(brokers []string) (*cluster, error) {
+	sort.Strings(brokers)
+	clusterKey := strings.Join(brokers, ",")
+
+	clusterLock.Lock()
+	defer clusterLock.Unlock()
+
+	if cluster, ok := clusters[clusterKey]; ok {
+		return cluster, nil
+	}
+
+	cluster, err := newCluster(brokers)
+	if err != nil {
+		return nil, err
+	}
+	clusters[clusterKey] = cluster
+	return clusters[clusterKey], nil
+}
+
+// newCluster creates a new cluster and returns it.
+func newCluster(brokers []string) (*cluster, error) {
+	brokerConf := kafka.NewBrokerConf("PortalMarshal")
+	broker, err := kafka.Dial(brokers, brokerConf)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &cluster{
+		quit:     new(int32),
+		rsteps:   new(int32),
+		broker:   broker,
+		producer: broker.Producer(kafka.NewProducerConf()),
+		topics:   make(map[string]int),
+		groups:   make(map[string]map[string]*topicState),
+		jitters:  make(chan time.Duration, 100),
+	}
+
+	// Do an initial metadata fetch, this will block a bit
+	err = c.refreshMetadata()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get metadata: %s", err)
+	}
+
+	// If there is no marshal topic, then we can't run. The admins must go create the topic
+	// before they can use this library. Please see the README.
+	c.partitions = c.getTopicPartitions(MarshalTopic)
+	if c.partitions == 0 {
+		return nil, errors.New("Marshalling topic not found. Please see the documentation.")
+	}
+
+	// Now we start a goroutine to start consuming each of the partitions in the marshal
+	// topic. Note that this doesn't handle increasing the partition count on that topic
+	// without stopping all consumers.
+	c.rationalizers.Add(c.partitions)
+	for id := 0; id < c.partitions; id++ {
+		go c.rationalize(id, c.kafkaConsumerChannel(id))
+	}
+
+	// A jitter calculator, just fills a channel with random numbers so that other
+	// people don't have to build their own random generator...
+	go func() {
+		rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+		for {
+			jitter := rnd.Intn(HeartbeatInterval/2) + (HeartbeatInterval / 2)
+			c.jitters <- time.Duration(jitter) * time.Second
+		}
+	}()
+
+	// Now start the metadata refreshing goroutine
+	go func() {
+		for !c.terminated() {
+			time.Sleep(<-c.jitters)
+			log.Infof("Refreshing topic metadata.")
+			c.refreshMetadata()
+
+			// See if the number of partitions in the marshal topic changed. This is bad if
+			// it happens, since it means we can no longer coordinate correctly.
+			if c.getTopicPartitions(MarshalTopic) != c.partitions {
+				log.Errorf("Marshal topic partition count changed. Terminating!")
+				c.terminate()
+			}
+		}
+	}()
+
+	// Wait for all rationalizers to come alive
+	log.Infof("Waiting for all rationalizers to come alive.")
```

Comment:
We probably want all of the logging in `cluster` to print which cluster we're talking about. Otherwise it's going to be hard to sort them out.

Maybe we could accept a human-friendly name for the cluster so we don't have to dump a list of hostnames here? That would mean changing the signature on `NewMarshaler` but it should probably take an options struct at this point anyway?